url53f943c5cb_Praznat_AGP_tgz-pJ8-AMath_TestingJ8/src/AMath/TestMarkets.java

```diff
@@ -0,0 +1,213 @@
+		resetMarketFunctions();
+		MktO timberMarket = ((MktO)doer1.myShire().getMarket(Misc.timber));
+		doer1.alterMillet(1000000);
+		((MktO)doer1.myShire().getMarket(Misc.timber)).placeBid(doer1, 10000);
+		doer1.MB.QuestStack.clear();
+		setupClanForWork(doer2.myShire(), doer2, Job.Lumberjacking);
+		doPursueUntilProduced(doer2, true);
+		affirm(timberMarket.getAskSz() > 0);
+	}
+	
+	public static void produceOne(Clan doer, int inG, int inN, int outG, int outN) {
+		final Labor labor = (Labor) doer.getJobActs()[0];
+		final int[] expIn = labor.expIn(doer);
+		for(int i = 1; expIn[i] != Misc.E; i++) {inG = expIn[i];}
+		final int[] expOut = labor.expOut(doer);
+		for(int i = 1; expOut[i] != Misc.E; i++) {outG = expOut[i];}
+		if(expOut[0] <= expIn[0]) {inN = 0; outN = 0;}
+		final Shire shire = doer.myShire();
+		final int rentInG = Assets.getRentGood(inG);
+		final int rentOutG = Assets.getRentGood(outG);
+		final int prevOutAskSz = shire.getMarket(outG).getAskSz();
+		final int prevInRentSz = rentInG >= 0 ? shire.getMarket(rentInG).getAskSz() : 0;
+		final int prevOutRentSz = rentOutG >= 0 ? shire.getMarket(rentOutG).getAskSz() : 0;
+		final MktAbstract inMarket = shire.getMarket(inG);
+		final boolean inputIsRental = inMarket instanceof RentMarket;
+		final int prevInBidSz = inMarket.getBidSz();
+		final int prevInAskSz = inputIsRental ? ((RentMarket)inMarket).numberUnrentedRemaining() : inMarket.getAskSz();
+		final int prevMyAsksForInG = ((MktO)inMarket).findNumberOf(doer, Entry.OFFERDIR);
+//		doPursueUntilInputsObtained(doer, true);
+//		affirm(inMarket.getAskSz() == prevInAskSz);
+//		affirm(((MktO)inMarket).findNumberOf(doer, Entry.OFFERDIR) == prevMyAsksForInG + inG);
+		doPursueUntilConsumed(doer, true);
+		if (inputIsRental) {affirm(((RentMarket)inMarket).isBestPlaceCorrect());}
+		if (inN > 0) {
+			affirm(inMarket.getBidSz() == prevInBidSz);
+			if (inputIsRental) {
+				affirm(((RentMarket)inMarket).numberUnrentedRemaining() == prevInAskSz - inN);
+			}	else {
+				affirm(inMarket.getAskSz() == prevInAskSz - inN);
+				affirm(((MktO)inMarket).findNumberOf(doer, Entry.OFFERDIR) == prevMyAsksForInG);
+			}
+		}
+		final int prevDoerOutN = outG < Misc.numAssets ? doer.getAssets(outG) : 0;
+		doPursueUntilProduced(doer, true);
+		//food is consumed so this will definitely need to change... meat is already being consumed during learn process
+		if (outG < Misc.numAssets && outG!=Misc.meat) affirm(doer.getAssets(outG) == prevDoerOutN + outN * (outG==Misc.millet ? FoodMarket.MILLETVAL: 1));
+		
+		if (!(shire.getMarket(outG) instanceof FoodMarket)) {affirm(shire.getMarket(outG).getAskSz() == prevOutAskSz + outN);}
+		if (rentInG >= 0) {
+			final RentMarket rMarket = (RentMarket) shire.getMarket(rentInG);
+			affirm(rMarket.getAskSz() == prevInRentSz - inN);
+			affirm(rMarket.bestOffer() <= shire.getMarket(inG).bestOffer());
+			affirm(rMarket.isBestPlaceCorrect());
+			rMarket.printBaikai();
+		}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 130:136, 137:143.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.